substrate/frame/election-provider-multi-block/src/unsigned/miner.rs

```diff
@@ -0,0 +1,811 @@
+
+//! # NPoS miner
+
+use crate::{
+	helpers,
+	types::{PageSize, Pagify},
+	unsigned::{pallet::Config as UnsignedConfig, Call},
+	verifier::FeasibilityError,
+	AssignmentOf, MinerSupportsOf, MinerVoterOf, Pallet as EPM, Snapshot,
+};
+
+use frame_election_provider_support::{
+	ElectionDataProvider, IndexAssignmentOf, NposSolution, NposSolver, PageIndex,
+	TryIntoBoundedSupports, Weight,
+};
+use frame_support::{ensure, traits::Get, BoundedVec};
+use scale_info::TypeInfo;
+use sp_npos_elections::{ElectionResult, ElectionScore, ExtendedBalance, Support};
+use sp_runtime::{offchain::storage::StorageValueRef, SaturatedConversion};
+use sp_std::{prelude::ToOwned, vec, vec::Vec};
+
+pub type TargetSnaphsotOf<T> =
+	BoundedVec<<T as Config>::AccountId, <T as Config>::TargetSnapshotPerBlock>;
+pub type VoterSnapshotPagedOf<T> = BoundedVec<
+	BoundedVec<MinerVoterOf<T>, <T as Config>::VoterSnapshotPerBlock>,
+	<T as Config>::Pages,
+>;
+
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub enum MinerError {
+	/// An internal error in the NPoS elections crate.
+	NposElections(sp_npos_elections::Error),
+	/// Snapshot data was unavailable.
+	SnapshotUnAvailable(SnapshotType),
+	/// An error from the election solver.
+	Solver,
+	/// The solution generated from the miner is not feasible.
+	Feasibility(FeasibilityError),
+	InvalidPage,
+	SubmissionFailed,
+	NotEnoughTargets,
+	DataProvider,
+}
+
+impl From<sp_npos_elections::Error> for MinerError {
+	fn from(e: sp_npos_elections::Error) -> Self {
+		MinerError::NposElections(e)
+	}
+}
+
+impl From<FeasibilityError> for MinerError {
+	fn from(e: FeasibilityError) -> Self {
+		MinerError::Feasibility(e)
+	}
+}
+
+impl From<SnapshotType> for MinerError {
+	fn from(typ: SnapshotType) -> Self {
+		MinerError::SnapshotUnAvailable(typ)
+	}
+}
+
+/// The type of the snapshot.
+///
+/// Used to express errors.
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub enum SnapshotType {
+	/// Voters at the given page missing.
+	Voters(PageIndex),
+	/// Targets are missing.
+	Targets,
+	// Desired targets are missing.
+	DesiredTargets,
+}
+
+/// Reports the trimming result of a mined solution
+#[derive(Debug, Clone, PartialEq)]
+pub struct TrimmingStatus {
+	weight: usize,
+	length: usize,
+}
+
+impl Default for TrimmingStatus {
+	fn default() -> Self {
+		Self { weight: 0, length: 0 }
+	}
+}
+
+use crate::PagedRawSolution;
+use codec::{EncodeLike, MaxEncodedLen};
+
+pub trait Config {
+	type AccountId: Ord + Clone + codec::Codec + core::fmt::Debug;
+
+	type Solution: codec::Codec
+		+ sp_std::fmt::Debug
+		+ Default
+		+ PartialEq
+		+ Eq
+		+ Clone
+		+ Sized
+		+ Ord
+		+ NposSolution
+		+ TypeInfo
+		+ EncodeLike
+		+ MaxEncodedLen;
+
+	type Solver: NposSolver<
+		AccountId = Self::AccountId,
+		Accuracy = <Self::Solution as NposSolution>::Accuracy,
+	>;
+
+	type Pages: Get<u32>;
+
+	type MaxVotesPerVoter: Get<u32>;
+	type MaxWinnersPerPage: Get<u32>;
+	type MaxBackersPerWinner: Get<u32>;
+
+	type VoterSnapshotPerBlock: Get<u32>;
```

Comment:
Shouldn't this be called `...PerPage` rather?
Maybe stupid question but line 146 with argument `all_voter_pages` made me think this way: it's a vector with max length equal to T::Pages so each entry in this vector should be "per page" therefore each entry should have at most N MinerVoterOf entities, where N is a cap on these voters "per page".